developer documentation for placeholders (#409)
caddyserver/website: src/docs/markdown/extending-caddy/placeholders.md

              "handle": [
                {
                  "body": "example",
                  "handler": "static_response",
                  "status_code": 200
                }
              ]
            }
          ]
        },
        "srv1": {
          "listen": [
            ":8081"
          ],
          "routes": [
            {
              "handle": [
                {
                  "body": "{env.HOST}",
                  "handler": "static_response",
                  "status_code": 200
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

Importantly, look at the `"body"` field in both `srv0` and `srv1`.

Since `srv0` used `{$HOST}`, the special environmental variable replacement with `$`, the value became `example`, as it was processed during Caddyfile parse time.

Since `srv1` used `{env.HOST}`, a normal placeholder, it was parsed as its own raw string value, `{env.HOST}`

Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.

## Implementing placeholder support

> You should not process placeholders when ummarshaling your Caddyfile. Instead, unmarshal the placeholders as strings in your configuration and evaluate them during either your module's execution or `Provision()` using a `caddy.Replacer`.

Reviewer: ```suggestion
You should not process placeholders when ummarshaling your Caddyfile. Instead, unmarshal the placeholders as strings in your configuration and evaluate them during either your module's execution (e.g. `ServeHTTP()` for HTTP handlers, `Match()` for matchers, etc.) or in the `Provision()` step, using a `caddy.Replacer`.
```

Author: reads well to me

---
title: "Placeholder Support"
---

# Placeholders

> In Caddy, placeholders are processed by each individual plugin themselves. They are not parsed at config time, but instead preserved and processed at runtime.

Reviewer 2: > They are not parsed at config time, but instead preserved and processed at runtime.

Technically, this depends if they're evaluated during Provision or later on. Maybe more specfic:

```suggestion
In Caddy, placeholders are processed by each individual plugin as needed; they do not automatically work everywhere.
```

This means that if you wish for your plugin to support placeholders, you must explicitly add support for them.

If you are not yet familiar with placeholders, start by reading [here](/docs/conventions#placeholders)!

## Placeholder Internals

> Internally, placeholders are simply a string in format `{foo.bar}` used as valid configuration values, which is later parsed at runtime.

Reviewer 2: ```suggestion
Placeholders are a string in the format `{foo.bar}` used as dynamic configuration values, which is later evaluated at runtime.
```

Placeholders-like strings which start with a dollar sign (`{$FOO}`), are evaulated at Caddyfile parse time, and do not need to be dealt with by your plugin. This is because these are not placeholders, but Caddyfile-specific [environmental variable substitution](/docs/caddyfile/concepts#environmental-variables), they just happen to share the `{}` syntax.

> It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`

Reviewer 2: ```suggestion
It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`.
```

---
title: "Placeholder Support"
---

# Placeholders

In Caddy, placeholders are processed by each individual plugin as needed; they do not automatically work everywhere.

This means that if you wish for your plugin to support placeholders, you must explicitly add support for them.

If you are not yet familiar with placeholders, start by reading [here](/docs/conventions#placeholders)!

> ## Placeholder Internals

Reviewer 2: ```suggestion
## Placeholder Overview
```

Internally, placeholders are simply a string in format `{foo.bar}` used as valid configuration values, which is later parsed at runtime.

Placeholders-like strings which start with a dollar sign (`{$FOO}`), are evaulated at Caddyfile parse time, and do not need to be dealt with by your plugin. This is because these are not placeholders, but Caddyfile-specific [environmental variable substitution](/docs/caddyfile/concepts#environment-variables), they just happen to share the `{}` syntax.

It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`

> As an example, see the following caddyfile:

Reviewer 2: ```suggestion
As an example, see the following Caddyfile:
```

          "routes": [
            {
              "handle": [
                {
                  "body": "example",
                  "handler": "static_response",
                  "status_code": 200
                }
              ]
            }
          ]
        },
        "srv1": {
          "listen": [
            ":8081"
          ],
          "routes": [
            {
              "handle": [
                {
                  "body": "{env.HOST}",
                  "handler": "static_response",
                  "status_code": 200
                }
              ]
            }
          ]
        }
      }
    }
  }
}
```

Importantly, look at the `"body"` field in both `srv0` and `srv1`.

Since `srv0` used `{$HOST}`, the special environmental variable replacement with `$`, the value became `example`, as it was processed during Caddyfile parse time.

Since `srv1` used `{env.HOST}`, a normal placeholder, it was parsed as its own raw string value, `{env.HOST}`

Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.

> ## How to use placeholders in your plugin

Reviewer 2: ```suggestion
## Implementing support for placeholders
```

Author: i chose `Implementing placeholder support` to better match the article title `implementing placeholders`

> #### Parse the raw placeholder value in your unmarshaler

Reviewer 2: I don't really think this section adds any value / says anything interesting. It can probably be removed...

Author: hmmm, my thought is that some people may be confused thinking they might need to use a special function to parse a placeholder.

if you don't think thats a problem, we can gut it

Reviewer 2: What if we remove the section but put what you're conveying in the parent section ("Implementing support for placeholders" as a heading I suggest):

> Placeholders are evaluated at runtime (not at Caddyfile parsing or config loading). When you are unmarshaling a Caddyfile, do not evaluate placeholders at that time. Placeholders may be evaluated during `Provision()` or during your module's execution.

Placeholders are not evaluated at Caddyfile parse time, and should be preserved for later use. They are used as their raw string values.

In other words, parsing a placeholder is no different from parsing any other string.

```go
func (g *Gizmo) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
	d.Next()
	if !d.Args(&g.Name) {
		// not enough args
		return d.ArgErr()
	}
}
```

> #### Evaluate the placeholder during Match or Serve

Reviewer 2: I would reframe this section and the next to be more general.

It boils down to whether a Replacer is already available or not. If one has already been made, as is the case with the HTTP handler chain, it can be obtained from the context. I think we can talk about it with that as the example, but less specific to "Match" or "ServeHTTP". The only real difference is how you get the Replacer: either from a context, or make a new one.

Author: do you think it should be reframed to be more general, or there be instead something two articles/examples showing

1. explaining how to grab a new replacer, and using that to resolve a placeholder
2. explaining how to create a replacer from a context.Context.

with a blurb saying that to use placeholder in code, pass the string into a replacer?

Reviewer 2: I think this one article is enough; mainly let's show examples of both in this one.

So show how to get it from the context, where it will often be stored, and also how to create a new one if there isn't one already available.

Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.

## Implementing placeholder support

You should not process placeholders when ummarshaling your Caddyfile. Instead, unmarshal the placeholders as strings in your configuration and evaluate them during either your module's execution or `Provision()` using a `caddy.Replacer`.

### Examples

> In this example, we are using a newly constructed replacer to process placeholders. It has access to placeholders such as `{env.HOST}`, but NOT `{http.request.uri}`

Reviewer: ```suggestion
In this example, we are using a newly constructed replacer to process placeholders. It has access to [global placeholders](/docs/conventions#placeholders) such as `{env.HOST}`, but NOT request placeholder such as `{http.request.uri}`
```

Author: i accepted this manually.
